pkg/clients/clients.go

```diff
@@ -1,57 +1,163 @@
+func (k *k8sClients) Update(ctx context.Context, obj client.Object) error {
+	return k.Update(ctx, obj)
 }
 
-func GetCachedDiscoveryClient() (discovery.CachedDiscoveryInterface, error) {
-	return config.ToDiscoveryClient()
+func (k *k8sClients) Get(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+	return k.Get(ctx, key, obj)
 }
 
-func HasResource(resource schema.GroupVersionResource) (bool, error) {
-	dclient, err := discovery.NewDiscoveryClientForConfig(RestConfig)
+func (k *k8sClients) Delete(ctx context.Context, obj client.Object) error {
+	return k.Delete(ctx, obj)
+}
+
+func (k *k8sClients) List(ctx context.Context, obj client.ObjectList, opts ...client.ListOption) error {
+	return k.List(ctx, obj, opts...)
+}
+
+func (k *k8sClients) Create(ctx context.Context, obj client.Object) error {
+	return k.Create(ctx, obj)
+}
+
+func (k *k8sClients) GetPodLogs(namespace, podName string, podLogOpts *v1.PodLogOptions) *restclient.Request {
+	return k.CoreV1().Pods(namespace).GetLogs(podName, podLogOpts)
+}
+
+func (k *k8sClients) GetNamespace(ctx context.Context, name string, opts metav1.GetOptions) (*v1.Namespace, error) {
+	return k.CoreV1().Namespaces().Get(ctx, name, opts)
+}
+
+func (k *k8sClients) GetSecret(ctx context.Context, namespace, name string, opts metav1.GetOptions) (*v1.Secret, error) {
+	return k.CoreV1().Secrets(namespace).Get(ctx, name, opts)
+}
+
+func (k *k8sClients) ClusterOperatorCreate(ctx context.Context, clusterOperator *configv1.ClusterOperator, opts metav1.CreateOptions) (*configv1.ClusterOperator, error) {
+	return k.ClusterOperators().Create(ctx, clusterOperator, opts)
+}
+
+func (k *k8sClients) ClusterOperatorGet(ctx context.Context, name string, opts metav1.GetOptions) (*configv1.ClusterOperator, error) {
+	return k.ClusterOperators().Get(ctx, name, opts)
+}
+
+func (k *k8sClients) ClusterOperatorUpdateStatus(ctx context.Context, clusterOperator *configv1.ClusterOperator, opts metav1.UpdateOptions) (*configv1.ClusterOperator, error) {
+	return k.ClusterOperators().UpdateStatus(ctx, clusterOperator, opts)
+}
+
+func (k *k8sClients) ClusterVersionGet(ctx context.Context, opts metav1.GetOptions) (result *configv1.ClusterVersion, err error) {
+	return k.ClusterVersions().Get(ctx, clusterVersionName, opts)
+}
+
+func (k *k8sClients) Invalidate() {
+	k.Invalidate()
+}
+
+func (k *k8sClients) ServerGroups() (*metav1.APIGroupList, error) {
+	return k.ServerGroups()
+}
+
+func (k *k8sClients) StatusUpdate(ctx context.Context, obj client.Object) error {
+	return k.Status().Update(ctx, obj)
+}
+
+func (k *k8sClients) CreateOrUpdate(ctx context.Context, obj client.Object, fn controllerutil.MutateFn) (controllerutil.OperationResult, error) {
+	return controllerruntime.CreateOrUpdate(ctx, k.Client, obj, fn)
+}
+
+func (k *k8sClients) HasResource(resource schema.GroupVersionResource) (bool, error) {
+	dclient, err := discovery.NewDiscoveryClientForConfig(k.restConfig)
```

Comment:
Wondering if the Discovery Client should be a struct member here. There are places in the code where it would be great to be able to mock it.

Reply:
i have looked into it. Currently we already have a CachedDiscoveryClient, so we need to verify that Discovery client does not have an interlocking interface. Since we are using it only in one case for now, i suggest we leave it as is ( we can always add it a define another method for it )